### PR TITLE
Share code for network modifications

### PR DESCRIPTION
**Pull request by Mathieu-Deharbe**

### CODE REVIEW

File: src/main/java/org/gridsuite/modification/server/modifications/VoltageLevelModification.java [hunk 1]
```diff
@@ -84,51 +85,73 @@ public void apply(Network network, ReportNode subReportNode) {
...
+            identifiableShortCircuitAdder.withIpMax(newIpMax);
+
+            //Convert to kA to report it like the user set it.
+            var oldIpMaxToReport = oldIpMax != null ? oldIpMax * 0.001 : null;
+            var newIpMaxToReport = newIpMax * 0.001;
+            reports.add(ModificationUtils.getInstance()
+                    .buildModificationReport(oldIpMaxToReport, newIpMaxToReport, "High short circuit current limit"));
+        } else if (oldIpMax != null) {
+            identifiableShortCircuitAdder.withIpMax(oldIpMax);
+        }
```
Comment (ghazwarhili) [inline src/main/java/org/gridsuite/modification/server/modifications/VoltageLevelModification.java:151]: to optimize and refacto this code I suggest something like 

```
     private void updateShortCircuitLimits(ValueWrapper ipMin, ValueWrapper ipMax, Double oldIpMin, Double oldIpMax, IdentifiableShortCircuitAdder identifiableShortCircuitAdder, List<ModificationReport> reports) {
        updateShortCircuitLimit(ipMin, oldIpMin, "Low short circuit current limit", identifiableShortCircuitAdder::withIpMin, reports);
        updateShortCircuitLimit(ipMax, oldIpMax, "High short circuit current limit", identifiableShortCircuitAdder::withIpMax, reports);
    }

    private void updateShortCircuitLimit(ValueWrapper newValueWrapper, Double oldValue, String limitDescription,
                                         Consumer<Double> setterMethod, List<ModificationReport> reports) {
        Double newValue = (newValueWrapper != null) ? newValueWrapper.getValue() : null;

        if (newValue != null) {
            setterMethod.accept(newValue);
            Double oldValueToReport = convertToKiloAmps(oldValue);
            Double newValueToReport = convertToKiloAmps(newValue);

            reports.add(ModificationUtils.getInstance().buildModificationReport(oldValueToReport, newValueToReport, limitDescription));
        } else if (oldValue != null) {
            setterMethod.accept(oldValue);
        }
    }

    private Double convertToKiloAmps(Double value) {
        return (value != null) ? value * 0.001 : null;
    }
```

File: src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModification.java [hunk 2]
```diff
@@ -62,50 +62,103 @@ protected void modifyCharacteristics(Branch<?> branch, BranchModificationInfos b
...
+    }
+
+    public static void modifyG(TwoWindingsTransformer transformer, AttributeModification<Double> modifG, ReportNode reportNode) {
+        if (modifG != null && modifG.getValue() != null) {
+            // convert reported value from siemens to microsiemens
+            if (reportNode != null) {
+                double oldMagnetizingConductanceToReport = transformer.getG() * Math.pow(10, 6);
+                double newMagnetizingConductanceToReport = modifG.getValue() * Math.pow(10, 6);
+                insertReportNode(reportNode, ModificationUtils.getInstance().buildModificationReportWithIndentation(oldMagnetizingConductanceToReport,
+                        newMagnetizingConductanceToReport, "Magnetizing conductance", 1));
```
Comment (ghazwarhili) [inline src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModification.java:142]: Use Constants for Keys

File: src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModification.java [hunk 3]
```diff
@@ -62,50 +62,103 @@ protected void modifyCharacteristics(Branch<?> branch, BranchModificationInfos b
...
+            }
+            transformer.setG(modifG.getValue());
         }
-        if (twoWindingsTransformerModificationInfos.getRatedU1() != null && twoWindingsTransformerModificationInfos.getRatedU1().getValue() != null) {
-            insertReportNode(characteristicsReporter, ModificationUtils.getInstance().buildModificationReportWithIndentation(twoWindingsTransformer.getRatedU1(),
-                    twoWindingsTransformerModificationInfos.getRatedU1().getValue(), "Rated Voltage (Side 1)", 1));
-            twoWindingsTransformer.setRatedU1(twoWindingsTransformerModificationInfos.getRatedU1().getValue());
+    }
+
+    public static void modifyBranchFields(TwoWindingsTransformer twt, AttributeModification<Double> modifR, AttributeModification<Double> modifX, ReportNode reportNode) {
```
Comment (EtienneLt) [inline src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModification.java:148]: change name and maybe separate in two methods

Comment (EtienneLt) [reply src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModification.java:148]: one for R and one for X

File: src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModification.java [hunk 4]
```diff
@@ -290,24 +347,31 @@ private void processRatioVoltageRegulation(RatioTapChangerModificationInfos rati
...
+                isModification ? ratioTapChanger::setTargetV
+                        : ratioTapChangerAdder::setTargetV,
+                isModification ? ratioTapChanger::getTargetV : () -> null,
+                targetV, "Target voltage", 2);
+
+        ReportNode targetDeadbandReportNode = ModificationUtils.getInstance().applyElementaryModificationsAndReturnReport(
+                isModification ? ratioTapChanger::setTargetDeadband
+                        : ratioTapChangerAdder::setTargetDeadband,
+                isModification ? ratioTapChanger::getTargetDeadband : () -> null,
+                targetDeadband, "Target deadband", 2);
```
Comment (EtienneLt) [inline src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModification.java:367]: the last value is often 2 maybe you could do a method with 2 as default value

File: src/test/java/org/gridsuite/modification/server/modifications/GeneratorByFormulaModificationTest.java [hunk 5]
```diff
@@ -299,7 +299,7 @@ protected void assertAfterNetworkModificationCreation() {
         assertNotNull(generatorStartup1);
         assertEquals(50, generator1.getTargetP(), 0);
         assertEquals(15, generatorStartup1.getMarginalCost(), 0);
-        assertEquals(55, generatorStartup1.getPlannedOutageRate(), 0);
+        assertEquals(2.5, generatorStartup1.getPlannedOutageRate(), 0);
```
Comment (EtienneLt) [inline src/test/java/org/gridsuite/modification/server/modifications/GeneratorByFormulaModificationTest.java:302]: why does it change ?

**Review by EtienneLt — CHANGES_REQUESTED** [review]
little comments

**Review by Mathieu-Deharbe — COMMENTED** [review]
Ok I renamed it to the whole `attributeModification`.

Many merge conflicts are solved in the last commits.

File: src/main/java/org/gridsuite/modification/server/dto/formula/equipmentfield/BatteryField.java [hunk 6]
```diff
@@ -36,14 +43,28 @@ public static Double getReferenceValue(Battery battery, String batteryField) {
 
     public static void setNewValue(Battery battery, String batteryField, Double newValue) {
         BatteryField field = BatteryField.valueOf(batteryField);
+        final AttributeModification<Double> modif = new AttributeModification<>(newValue, OperationType.SET);
```
[the remark this answers is not in the document]
Comment (Mathieu-Deharbe) [reply src/main/java/org/gridsuite/modification/server/dto/formula/equipmentfield/BatteryField.java]: Ok I renamed it to the whole attributeModification.

File: src/main/java/org/gridsuite/modification/server/dto/formula/equipmentfield/TwoWindingsTransformerField.java [hunk 7]
```diff
@@ -48,23 +51,32 @@ public static void setNewValue(TwoWindingsTransformer transformer, String twoWin
         TwoWindingsTransformerField field = TwoWindingsTransformerField.valueOf(twoWindingsTransformerField);
         final PhaseTapChanger phaseTapChanger = transformer.getPhaseTapChanger();
         final RatioTapChanger ratioTapChanger = transformer.getRatioTapChanger();
+        final AttributeModification<Double> attrModif = new AttributeModification<>(newValue, OperationType.SET);
```
[the remark this answers is not in the document]
Comment (Mathieu-Deharbe) [reply src/main/java/org/gridsuite/modification/server/dto/formula/equipmentfield/TwoWindingsTransformerField.java]: Ok I renamed it to the whole attributeModification.

**Review by basseche — APPROVED** [review]

File: src/main/java/org/gridsuite/modification/server/modifications/GeneratorModification.java [hunk 8]
```diff
@@ -418,9 +456,25 @@ private void modifyGeneratorSetpointsAttributes(GeneratorModificationInfos modif
         modifyGeneratorActivePowerControlAttributes(modificationInfos, generator, subReportNode, subReporterSetpoints);
     }
 
+    public static ReportNode modifyTargetQ(Generator generator, AttributeModification<Double> modifTargetQ) {
+        ReportNode reportReactivePower = null;
+        if (modifTargetQ != null) {
+            if (modifTargetQ.getOp() == OperationType.SET) {
+                reportReactivePower = ModificationUtils.getInstance().applyElementaryModificationsAndReturnReport(generator::setTargetQ, generator::getTargetQ, modifTargetQ, "Reactive power");
```
Comment (EtienneLt) [inline src/main/java/org/gridsuite/modification/server/modifications/GeneratorModification.java:463]: should it be specified that it is **target** reactive power or setpoint ?

**Review by EtienneLt — COMMENTED** [review]
few more comments and it is good for me

**Review by EtienneLt — APPROVED** [review]

**Pull request merged**